### DISCUSSION TITLE
Does a simple 'repeat last command' (for _any_ command) feature exist?

### DISCUSSION THREAD

**Discussion by johnnyutahh** [helix-editor/helix#8311] [Q&A]
Is there a `repeat last deletion` command/feature?  If not, why not?

I see lots of "repeat last" conversations (below), but have yet to see a specific `repeat last deletion` discussion/reference.

https://github.com/search?q=repo%3Ahelix-editor%2Fhelix+repeat+last&type=issues
https://github.com/search?q=repo%3Ahelix-editor%2Fhelix+repeat+last&type=discussions

**Comment by johnnyutahh** [comment]
Digging a little further...

https://github.com/helix-editor/helix/blob/master/book/src/keymap.md has a `N/A` in the `Command` column for the `. / Repeat last insert` row. That seems weird.

A little code searching (in session below) reveals a `repeat_last_motion()` function (I think) but nothing like a "repeat last insert". Also seems strange.

If someone can spell out where the "repeat last insertion/change" stuff is happening, I _might_ be able to look further, maybe even generate an _"add repeat last delete/deletion"_ PR for `.`... or some such. (I'm not good at Rust yet, so it might take a while.)

```
$ git checkout -b 23.05
Switched to a new branch '23.05'
$
$ git remote show origin | grep URL; git rev-parse HEAD
  Fetch URL: https://github.com/helix-editor/helix
  Push  URL: https://github.com/helix-editor/helix
7fa5f341e931f9bdb4058e51ebf77276e9b82242
$
$ ag 'repeat.*last.*insert'
runtime/tutor
709: Type . to repeat the last insert command.
736: * Type . to repeat the last insertion.

book/src/keymap.md
77:| `.`         | Repeat last insert                                                   | N/A                       |
$
$ ag repeat_last
helix-view/src/editor.rs
1073:    pub fn repeat_last_motion(&mut self, count: usize) {

book/src/keymap.md
51:| `Alt-.`               | Repeat last motion (`f`, `t` or `m`)               | `repeat_last_motion`        |

helix-term/src/keymap/default.rs
20:        "A-." =>  repeat_last_motion,

helix-term/src/commands.rs
268:        repeat_last_motion, "Repeat last motion",
1462:fn repeat_last_motion(cx: &mut Context) {
1463:    cx.editor.repeat_last_motion(cx.count())
$
```

**Comment by johnnyutahh** [comment]
Related issue/discussion here (for us historically-`vim`-using users):
https://github.com/LGUG2Z/helix-vim/issues/22

**Comment by johnnyutahh** [comment]
I'd also be happy to map an additional/separate (from `.`) normal-mode command/keystroke to "repeat last deletion"... but there's no such mappable command/function/whatever in `helix-term/src/commands.rs:impl MappableCommand {...}`.

So again, I'm wondering: how does `.` operate today? It seems to not be plugged in with `helix-term/src/commands.rs:impl MappableCommand {...}`. Maybe the `repeat` calls/code/stuff in `fn paste_impl()` manages this magic?  (My Rust-programming capability sucks right now, so I can't yet decipher.)

**Comment by gabydd** [comment]
Repeat is handled here: https://github.com/helix-editor/helix/blob/master/helix-term/src/ui/editor.rs#L895-L929 it would probably be nice if it was moved to a command I wonder how you would see a repeat last delete implemented/what it would do because the delete command just deletes whatever the current selection is which doesn't translate well to being repeated because how do you recreate a selection that was deleted?

**Comment by johnnyutahh** [comment]
> Repeat is handled here: https://github.com/helix-editor/helix/blob/master/helix-term/src/ui/editor.rs#L895-L929 it would probably be nice if it was moved to a command

Very helpful, many thanks @gabydd.

> I wonder how you would see a repeat last delete implemented/what it would do because the delete command just deletes whatever the current selection is which doesn't translate well to being repeated because how do you recreate a selection that was deleted?

The goal is to make a _new_ selection, repeating the _last_ "command + selection" (or "selection + command"?) that was made. Just like vim does. Seems feasible to me. Maybe I'm missing something. Or maybe I misinterpreted what you're writing above?

eg: issuing `wd` (in `hx`) deletes the "current" word (I think). I propose that successively issuing/pressing `.` will delete the next word. (Just like `dw` and a successive `.` in `vim`.)

Of course, I might be mixing `hx`-vs-`vim` concepts here; I'm a long-time `vim` user but new to `hx`. Regardless: the `.` "repeat last command" feature is my most-used command in `vim`. I probably use it for delete commands more than any other. How in the world this did not get implemented in `hx` is... a _major_ mystery to me. Maybe I'm one of the few if only `vim` users that runs `dd` (delete current line) or `dw` (delete current word) or `d}` (delete next/current paragraph) `vim` commands followed by numerous `.` to successively-and-"visually" delete. Super powerful in my world. I'm quite surprised this has not yet been included in `hx` workflows. I suspect there's some "secret" or "trick" or feature or workflow I do not yet know.

**Reply by gabydd** [reply]
Ah so you want something that repeats the last selection and then the last action I see yeah if there was a repeat last action or delete command you could pair that with repeat last motion to achieve this.

For something like visual deleting a bunch of lines of text there are two options:
1. press `x` until you have selected all the lines you want to delete and then press ``d``
2. enter extend mode with `v` and then use `j` or `k` to extend the selection up or down to select the lines you want to delete and then at the end you can press ``Xd`` to delete all the selected lines.

For visually deleting some words you can enter extend mode with ``v`` then press``w`` till you have selected all the words you want to delete and then press ``d`` to delete them.

(This is probably not what you want for the cases you described so I'll move it to the end) In lots of cases where people would use repeat you will probably want to use multiple selections usually using ``s`` to make a subselection of your selection or ``Alt-s`` to split your selection on newlines.

**Comment by johnnyutahh** [comment]
I simply want to repeat the last command (or command "set"). It's this simple: run a normal-mode command and then repeat said command... REGARDLESS of the `hx`-normal-mode command... with `.` (or any other key/map). Just like `vim` does.

Maybe it would help if some ex-`vim` users jumped in to amplify this point?

The (kind of long and complex) paragraph below feels... far from simple.  :-/  It's a bit hard for me to follow. However, I want to give a BIG THANK YOU for helping me with this @gabydd.  🤓

> Ah so you want something that repeats the last selection and then the last action I see yeah if there was a repeat last action or delete command you could pair that with repeat last motion to achieve this. In lots of cases where people would use repeat you will probably want to use multiple selections usually using `s` to make a subselection of your selection or `Alt-s` to split your selection on newlines. for something like visual deleting a bunch of lines of text there are two options either you can use `x` to select a bunch of lines and then delete or you can enter extend mode with `v` and then use j or k to select the lines you want to delete and then at the end you can press `Xd` to delete all the selected lines. for visually deleting some words you can enter extend mode again with `v`then do all the `w`s you need and then press once you are done selecting all your words.

**Comment by johnnyutahh** [comment]
I can also see how the normal-and-visual-`hx` modes to "select a bunch of things/lines/words/paragphs and _then_ delete" might replace `vim`'s `.` (repeat)... for deletions.

But...

1. why can't I/we have both (`vim`'s `.` repeat _and_ `hx`'s current select-better-paradigm)?
1. I want a command repeat that repeats _anything_, not just deletes or inserts. But ANY last normal-mode command (...just like `vim` does).
1. is it hard to implement a dumb, simple `vim` `.` repeat (in `hx`)? if not... why not just do it?

**Reply by gabydd** [reply]
you can track the issue for this here: https://github.com/helix-editor/helix/issues/3869

why this has not been implemented is for one nobody has tried to do it yet. As well this is not something that the maintainers are interested in and as they are all doing this in their free time as volunteers they are mostly going to work on things they believe are priorities and are interested in, so they can't be expected to implement. I am not sure how difficult it would be to implement this but feel free to attempt it

**Reply by johnnyutahh** [reply]
> you can track the issue for this here: #3869

Got it. Super helpful, thank you!
 
> why this has not been implemented is for one nobody has tried to do it yet. As well this is not something that the maintainers are interested in and as they are all doing this in their free time as volunteers they are mostly going to work on things they believe are priorities and are interested in, so they can't be expected to implement. I am not sure how difficult it would be to implement this but feel free to attempt it

This all makes perfect sense.

These points, along with the repeat-code reference (https://github.com/helix-editor/helix/blob/master/helix-term/src/ui/editor.rs#L895-L929) have all provided excellent clarity on this matter. Thanks again @gabydd!

**Comment by johnnyutahh** [comment]
(I edited the title of this discussion to reflect the broader scope.)

**Comment by pascalkuthe** [comment]
All of this seems like a macro.

In vim repeat makes more sense because selection and action are coupled (dw is one command) but in helix it's not so it's unclear how many commands should ne recorded/repeated.


We already have macros that record a certain key sequences and then you can repeat them. I don't think we would add anything beyond that. 

In general helix is based around multi cursors so instead of repeating something often you would select everything you would want to delete at once only press `d` once (for example enter visual mode, select all words you want with d and then finally delete with d)

**Comment by johnnyutahh** [comment]
**Question**: Is it possible (for me or others) to add a "repeat last *_deletion_*" (in say `helix-term/src/commands.rs:impl MappableCommand {...}`)... that I can then map to another normal-mode command/keystroke)?

There seems to be "repeat last movement" and "repeat last insert/change" functionality. 

replies to and questions surrounding @pascalkuthe's above points:

> In vim repeat makes more sense because selection and action are coupled (dw is one command) but in helix it's not so it's unclear how many commands should ne recorded/repeated.

Ah. This seems like the core of the challenge. Got it.

> We already have macros that record a certain key sequences and then you can repeat them. I don't think we would add anything beyond that.

Right. If I understand this correctly, that means making a diff macro for each command sequence. That's in some ways helpful, thanks. However, it's not a general, single "repeat last command" manuever/command/keystroke that I can reuse in many cases, if I understand things correctly...?

> In general helix is based around multi cursors so instead of repeating something often you would select everything you would want to delete at once only press `d` once (for example enter visual mode, select all words you want with d and then finally delete with d)

Yep, that makes sense, and this user-interaction "philosophy" seems clear, and seems like a reasonable alternative to my old `vim` workflow.

I'd still like to see if a "repeat last deletion" mechanism is possible--per my first question, above.

**Comment by davidhooey** [comment]
I too find it frustrating to not be able to repeat the last delete of a line or word. I often use `dd` in vim and then `.` to repeat the line deletions over and over. In Helix I issue a `xd` to select the line and delete it, I then expect the `.` to repeat `xd` again. Or even issue the `.` once to see the repeated select `x` so the user can see the new selection to ensure it is what is needed, then issue `.` a second time to confirm the delete. A double tap of `.` is much faster than repeating `xd`.

These is one of the main issues that has prevented me from moving from nvim to helix.

**Reply by kirawi** [reply]
See the second paragraph of https://github.com/helix-editor/helix/issues/501#issuecomment-1869920017
